app/services/analytics_service.rb

```diff
@@ -71,8 +72,15 @@ def stats_grouped_by_day
   attr_reader :user_or_org, :start_date, :end_date, :article_data, :reaction_data, :comment_data, :follow_data, :page_view_data
 
   def load_data
-    @article_data = Article.where("#{user_or_org.class.name.downcase}_id" => user_or_org.id, published: true)
-    article_ids = @article_data.pluck(:id)
+    if @single_article_id
+      @article_data = Article.where(id: @single_article_id, "#{user_or_org.class.name.downcase}_id" => user_or_org.id)
```

Comment:
A couple of things:

you're missing the "published: true" check here and if you use [Article.exists?](https://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-exists-3F) instead of where you avoid loading an article just to check  if it's blank. You can ask the DB "do you have an article matching these params?"

Reply:
Hmm, so something like:
```ruby
raise UnauthorizedError if @single_article_id && !Article.exists(id: @single_article_id, published: true, "#{user_or_org.class.name.downcase}_id" => user_or_org.id)

    if @single_article_id
      @article_data = Article.where(id: @single_article_id, published: true, "#{user_or_org.class.name.downcase}_id" => user_or_org.id)
      article_ids = @single_article_id
    else
      @article_data = Article.where("#{user_or_org.class.name.downcase}_id" => user_or_org.id, published: true)
      article_ids = @article_data.pluck(:id)
    end
```

Reply:
Also good catch on the `published: true` 👍

Reply:
@Zhao-Andy I think your solution was better, because now we have two queries. Just add `published: true` and forget about the other suggestion :D